File: M_Simulator_FEM_2D/08_M_Simulator_FEM_2D/MATLAB/Solver_08_M_Simulator_FEM_2D.m
%% - Elastic FEM Mechanics Simulator For Quadrilatoral Elements Only 2-D
%Barnaby Fryer 20.04.17

clc
clear all
close all

%% - Input Data
[Gen, Plotting, Pos] = Input_FEM();

%% - Force vector, determine where to apply forces here
[F] = Build_Force(Gen);

%% - Stiffness Matrix
[k_global] = Global_Stiffness_Builder(Gen);

%% - Boundary Conditions
%Apply fixes to global matrix
[F,k_global] = Fix_x(F,k_global,Gen.nodesx);
[F,k_global] = Fix_y(F,k_global,Gen.nodesy);

%% - Solve
Pos.dx = k_global\F;                                            %[2*Nn,1]

%% - Process displacements
%Displacement in x-direction
Pos.u = Pos.dx(1:2:length(Pos.dx),1);                           %[Nn,1]
%Displacement in y-direction
Pos.v = Pos.dx(2:2:length(Pos.dx),1);                           %[Nn,1]

%Displacement in x-direction reshaped
Pos.du = reshape(Pos.u,Gen.Nx+1,Gen.Ny+1);                      %[Nx+1,Ny+1]
%Displacement in y-direction reshaped
Pos.dv = reshape(Pos.v,Gen.Nx+1,Gen.Ny+1);                      %[Nx+1,Ny+1]

%New node locations
Pos.x_new = Pos.x + Pos.du;                                     %[Nx+1,Ny+1]
Pos.y_new = Pos.y + Pos.dv;                                     %[Nx+1,Ny+1]

%% - Solve for stresses
%Solve for stresses
[Sig] = Solve_Stresses(Gen,Pos);                                %[Nn,3]

%% - Plot
Plotting_sim(F,Gen,Plotting,Pos,Sig);

File: M_Simulator_FEM_2D/08_M_Simulator_FEM_2D/Python/Solver_08_M_Simulator_FEM_2D.py
# - About
# Elastic FEM Mechanics Simulator For Quadrilatoral Elements Only 2-D

# Barnaby Fryer, 2026

# =============================================================================
# Import libraries
# =============================================================================

import numpy as np

from src.build_force import build_force
from src.input_data import input_data
from src.global_stiffness_builder import global_stiffness_builder
from src.fix_x import fix_x
from src.fix_y import fix_y
from src.solve_stresses import solve_stresses
from src.plotting_sim import plotting_sim

# =============================================================================
# Inputs
# =============================================================================

Gen, Plotting, Pos = input_data()

# =============================================================================
# Force vector, determine where to apply forces here
# =============================================================================

f = build_force(Gen)

# =============================================================================
# Build stiffness matrix
# =============================================================================

k_global = global_stiffness_builder(Gen)

# =============================================================================
# Boundary conditions
# =============================================================================

#Apply fixes to global matrix
f, k_global = fix_x(f, k_global, Gen["nodes_left"])
f, k_global = fix_y(f, k_global, Gen["nodes_bottom"])

# =============================================================================
# Solve
# =============================================================================

Pos["dx"] = np.linalg.solve(k_global, f)

# =============================================================================
# Process displacements
# =============================================================================

# Displacement in x-direction (u)
Pos["u"] = Pos["dx"][0::2]

# Displacement in y-direction (v)
Pos["v"] = Pos["dx"][1::2]

# Reshape to grid form
Pos["du"] = Pos["u"].reshape((Gen["Ny"] + 1, Gen["Nx"] + 1))
Pos["dv"] = Pos["v"].reshape((Gen["Ny"] + 1, Gen["Nx"] + 1))

# New node locations
Pos["x_new"] = Pos["x"] + Pos["du"]
Pos["y_new"] = Pos["y"] + Pos["dv"]

# =============================================================================
# Solve for stresses and volumetric strain
# =============================================================================

Sigma, e_vol = solve_stresses(Gen, Pos)

# =============================================================================
# Plotting
# =============================================================================

plotting_sim(f, Gen, Plotting, Pos, Sigma, e_vol)

